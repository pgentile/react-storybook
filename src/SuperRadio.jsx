import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';

import styles from './SuperRadio.scss';


class SuperRadio extends React.PureComponent {

  static propTypes = {
    label: PropTypes.string.isRequired,
    icon: PropTypes.object.isRequired,
    description: PropTypes.string,
    checked: PropTypes.bool,
    onChange: PropTypes.func,
    forwardedRef: PropTypes.any,
  };

  static counter = 0;

  constructor(props) {
    super(props);

    SuperRadio.counter++;

    this.state = {
      id: `super-radio-${SuperRadio.counter}`,
    };
  }

  render() {
    const { label, icon, description, checked, onChange, forwardedRef } = this.props;
    const { id } = this.state;

    return (
      <div className={styles.superRadio}>
        <div className={styles.superRadio__radioElement}>
          <input type="radio" id={id} checked={checked} onChange={onChange} ref={forwardedRef} />
        </div>
        <div className={styles.superRadio__icon}>
          <label htmlFor={id}>
            <FontAwesomeIcon icon={icon} size="2x" />
          </label>
        </div>
        <div className={styles.superRadio__labelElement}>
          <p className={styles.superRadio__labelTitle}>
            <label htmlFor={id}>{label}</label>
          </p>
          {description && <p className={styles.superRadio__labelDescription}>{description}</p>}
        </div>
      </div>
    );
  }

}


export default forwardRef((props, ref) => {
  return (
    <SuperRadio forwardedRef={ref} {...props} />
  );
});
