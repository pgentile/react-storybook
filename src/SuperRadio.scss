$baseBackgroundColor: lightgreen;
$baseBorderColor: darkgray;
$margin: 10px;


.super-radio {
  margin: $margin;
  background-color: $baseBackgroundColor;
  border-radius: 7px;
  border: 2px $baseBorderColor solid;
  display: flex;
  flex-direction: row;
  justify-content: left;
  align-items: center;

  &_checked {
    background-color: darken($baseBackgroundColor, 10%);
    box-shadow: 0 3px 10px lighten($baseBorderColor, 10%);
    border-color: darken($baseBorderColor, 20%);
  }

  &__radio-element {
    margin: auto 0 auto $margin;
  }

  &__icon {
    margin: auto 0 auto $margin;
  }

  &__label-element {
    margin: auto 0;
  }

  &__label-title {
    margin: $margin;
    font-weight: bold;
  }

  &__label-description {
    margin: $margin;
  }

}

