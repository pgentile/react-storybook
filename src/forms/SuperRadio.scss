.super-radio {
  $base-margin: 0.5rem;
  $description-margin: -$base-margin;

  &__main-container {
    padding: $base-margin;
    display: flex;
    flex-direction: row;
    justify-content: left;
    align-items: center;
  }

  &__radio-element {
    margin-right: $base-margin;
    flex-grow: 0;
  }

  &__icon {
    margin-right: $base-margin;
    flex-grow: 0;
  }

  &__description {
    // top | right | bottom | left
    margin: $description-margin 0 $description-margin $description-margin;
    flex-grow: 1;
  }

  &__description-title {
    margin: $base-margin;
    margin-right: 0;
    font-weight: bold;
    font-size: 1.1rem;
  }

  &__description-description {
    margin: $base-margin;
    margin-left: $base-margin;
  }

  &__help {
    // top | right | bottom | left
    margin: $description-margin 0 $description-margin $description-margin;
    flex-grow: 0;
  }

  &__help-link {
    cursor: pointer;
  }
}
