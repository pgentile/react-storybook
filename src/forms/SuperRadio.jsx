import React, { createRef } from "react";
import PropTypes from "prop-types";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faQuestionCircle } from "@fortawesome/free-solid-svg-icons";

import ExpandableCard from "../ExpandableCard";

import "./SuperRadio.scss";
import bemModifiers from "../utils/bemModifiers";

export default class SuperRadio extends React.PureComponent {
  static propTypes = {
    label: PropTypes.node.isRequired,
    icon: PropTypes.object,
    description: PropTypes.node,
    help: PropTypes.node,
    checked: PropTypes.bool,
    onChange: PropTypes.func,
  };

  static counter = 0;

  state = {
    showHelp: false,
  };

  inputRef = createRef();
  inputId = `super-radio-${++SuperRadio.counter}`;

  selectRadio = (event) => {
    event.stopPropagation();

    const input = this.inputRef.current;
    if (input && !input.checked) {
      input.focus();
      input.click();
    }
  };

  toggleHelp = (event) => {
    event.stopPropagation();

    this.setState((prevState) => {
      return {
        showHelp: !prevState.showHelp,
      };
    });
  };

  stopPropagation = (event) => {
    event.stopPropagation();
  };

  render() {
    const { label, icon, description, help, checked, onChange } = this.props;
    const { showHelp } = this.state;

    const realClassName = bemModifiers("super-radio", {
      checked,
    });

    return (
      <ExpandableCard className={realClassName} expandableContent={help} expanded={showHelp} hasBorder layer="flat">
        <div className="super-radio__main-container" onClick={this.selectRadio}>
          <div className="super-radio__radio-element">
            <input
              ref={this.inputRef}
              type="radio"
              id={this.inputId}
              checked={checked}
              onChange={onChange}
              onClick={this.stopPropagation}
            />
          </div>
          {icon && (
            <div className="super-radio__icon">
              <label htmlFor={this.inputId} onClick={this.stopPropagation}>
                <FontAwesomeIcon icon={icon} size="3x" />
              </label>
            </div>
          )}
          <div className="super-radio__description">
            <p className="super-radio__description-title">
              <label htmlFor={this.inputId}>{label}</label>
            </p>
            {description && <p className="super-radio__description-description">{description}</p>}
          </div>
          {help && (
            <div className="super-radio__help">
              <a className="super-radio__help-link" onClick={this.toggleHelp}>
                <FontAwesomeIcon className="super-radio__help-icon" icon={faQuestionCircle} size="1x" />
              </a>
            </div>
          )}
        </div>
      </ExpandableCard>
    );
  }
}
