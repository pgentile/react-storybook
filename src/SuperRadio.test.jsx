import React from "react";
import renderer from "react-test-renderer";

import { faCoffee } from "@fortawesome/free-solid-svg-icons";

import SuperRadio from "./SuperRadio";

test("Super radio", () => {
  const component = renderer.create(<SuperRadio label="Label" description="This is my radio button" icon={faCoffee} />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
