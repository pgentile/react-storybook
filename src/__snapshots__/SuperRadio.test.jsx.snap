// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Super radio 1`] = `
<div
  className="super-radio "
  onClick={[Function]}
>
  <div
    className={
      Object {
        "modifier": [Function],
        "toString": [Function],
      }
    }
  >
    <input
      checked={undefined}
      id="super-radio-1"
      onChange={undefined}
      type="radio"
    />
  </div>
  <div
    className={
      Object {
        "modifier": [Function],
        "toString": [Function],
      }
    }
  >
    <label
      htmlFor="super-radio-1"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-coffee fa-w-20 fa-3x "
        data-icon="coffee"
        data-prefix="fas"
        role="img"
        style={Object {}}
        viewBox="0 0 640 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M192 384h192c53 0 96-43 96-96h32c70.6 0 128-57.4 128-128S582.6 32 512 32H120c-13.3 0-24 10.7-24 24v232c0 53 43 96 96 96zM512 96c35.3 0 64 28.7 64 64s-28.7 64-64 64h-32V96h32zm47.7 384H48.3c-47.6 0-61-64-36-64h583.3c25 0 11.8 64-35.9 64z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </label>
  </div>
  <div
    className={
      Object {
        "modifier": [Function],
        "toString": [Function],
      }
    }
  >
    <p
      className={
        Object {
          "modifier": [Function],
          "toString": [Function],
        }
      }
    >
      <label
        htmlFor="super-radio-1"
      >
        Label
      </label>
    </p>
    <p
      className={
        Object {
          "modifier": [Function],
          "toString": [Function],
        }
      }
    >
      This is my radio button
    </p>
  </div>
</div>
`;
